platform/include/platform/mbed_retarget.h

```diff
@@ -63,17 +64,36 @@ typedef unsigned int  gid_t;    ///< Group ID
 /* Flags for open() and fcntl(GETFL/SETFL)
  * At present, fcntl only supports reading and writing O_NONBLOCK
  */
+#ifndef O_RDONLY
```

Comment:
I am trying to find a reasoning we had it previously as undef. Didn't we have fcntl in one of the toolchains (I bet it was IAR back then) ? If this is tested with both toolchains we support, it should be good.

Avoiding redefinition errors (either this PR or as it was would result with the same effect), however the previous one ignored the values set by elsewhere and forcing our retarget value (undef + set). This PR is changing it to (ifndef then set).

@kjbracey-arm Would you know if the proposed way is acceptable ?

Reply:
I think ifndef is better for errno because it maintains binary compatibility with the C library. There will be some C library functions setting errno (but not many in this embedded environment - usually it's the porting layer), and they would be using the values from the toolchain header.

I'm not sure what the original justification for the forced undefine was. Consistency of values between toolchains?

Comment says this, but it's not totally illuminating:

    * Note also that ARMCC errno.h defines some symbol values differently from
    * the GCC_ARM/IAR/standard POSIX definitions. The definitions guard against
    * this and future changes by changing the symbol definition as shown below.

The implication being that the values below were matching GNU+IAR?

Switching values now (for ARMC only?) would potentially introduce binary compatibility issues with other binary blobs. But I doubt that would matter. The sort of things using binary blobs don't use these values - Wifi drivers would be using NSAPI error codes.

So I don't have a problem with this change, conceptually.

The main issue is that if doing ifdef it really must make sure it has included the toolchain file that would define these to get the same result every time.

Otherwise

    #include "mbed_retarget.h"

and

    #include <sys/something>
    #include "mbed_retarget.h"

might give you two different values in different compilation units.

Reply:
Thanks for commenting on this Kevin. 

Most of the values below match GCC's newlib values but not all of them; some are different. It is what motivated the change. For example, `EADDRINUSE` is equal to `112` on Mbed OS but is set to `98` in newlib. 

This has caused us issues when implementing a POSIX compatible socket layer on top of Mbed OS as the values between the system header and mbed header were not matching. 

The inclusion of `errno.h` and `fcntl.h` should cover us as defines are reused from here. 

The last difficulty is to find if a value defined in this file doesn't collide with another errno value defined in the system header. I think we can use a `constexpr` and `static_assert` (maybe just in the cpp file to not impact build time) to validates that these errno values are distinct:

```c++
constexpr bool validate_errno_values() { 
    constexpr int values[] { 
        EPERM,
        ENOENT,
        ESRCH
        /* other errno values ... */
    };

    constexpr int count = sizeof(values) / sizeof(values[0]);

    for (int i = 0; i < count; ++i) { 
        for (int j = i + 1; j < count; ++j) { 
            if (i == j) { 
                continue;
            }
            if (values[i] == values[j]) { 
                return false;
            }
        }
    }

    return true;
}

static_assert(validate_errno_values() == true, "Not all errno values are unique");
```
 
@tymoteuszblochmobica Have you validated the change with ARMCC ?

Reply:
Cute, but you don't need to go that far. Just make a `switch` statement listing each value. That will give a compile error for duplicates.

(EAGAIN and EWOULDBLOCK are allowed to be duplicates, others aren't, I believe).

Reply:
Aha, I completely forgot about the switch statement 🤦 . That should work too and be faster.